src/Printer_CartridgeInfo.php

```diff
@@ -115,4 +115,147 @@ public function showForPrinter(Printer $printer)
+                'table'             => self::getTable(),
+                'field'             => "_virtual_toner_{$color_key}_percent",
+                'name'              => sprintf(__('%s toner percentage'), $color_name),
+                'datatype'          => 'specific',
+                'massiveaction'     => false,
+                'nosearch'          => true,
+                'joinparams'        => [
+                    'jointype' => 'child'
+                ],
+                'additionalfields'  => ['property', 'value'],
+                'forcegroupby'      => true,
+                'aggregate'         => true,
+                'searchtype'        => ['contains'],
+                'nosort'            => true
+            ];
+        };
+
+        $tab[] = $create_toner_percent_option(1400, 'black', __('Black'));
+        $tab[] = $create_toner_percent_option(1401, 'cyan', __('Cyan'));
+        $tab[] = $create_toner_percent_option(1402, 'cyanlight', __('Light cyan'));
+        $tab[] = $create_toner_percent_option(1403, 'magenta', __('Magenta'));
+        $tab[] = $create_toner_percent_option(1404, 'magentalight', __('Light magenta'));
+        $tab[] = $create_toner_percent_option(1405, 'yellow', __('Yellow'));
+        $tab[] = $create_toner_percent_option(1406, 'grey', __('Grey'));
+        $tab[] = $create_toner_percent_option(1407, 'darkgrey', __('Dark grey'));
+
+        return $tab;
+    }
+
+    public static function getSpecificValueToSelect($field, $name = '', $values = '', array $options = [])
+    {
+        return parent::getSpecificValueToSelect($field, $name, $values, $options);
+    }
+
+    private static function getProgressColorsForColor($color)
+    {
+        $fg_transparency_hex = '80';
+        switch ($color) {
+            case 'cyan':
+            case 'light_cyan':
+                return [
+                    'fg' => '#00ffff' . $fg_transparency_hex,
+                    'text' => 'inherit'
+                ];
+            case 'magenta':
+            case 'light_magenta':
+                return [
+                    'fg' => '#ff00ff' . $fg_transparency_hex,
+                    'text' => 'inherit'
+                ];
+            case 'yellow':
+                return [
+                    'fg' => '#ffff00' . $fg_transparency_hex,
+                    'text' => 'inherit'
+                ];
+            case 'black':
+            case 'grey':
+            case 'darkgrey':
+                return [
+                    'fg' => '#303030' . $fg_transparency_hex,
+                    'text' => 'inherit'
+                ];
+        }
+
+        return null;
+    }
+
+    public static function getSpecificValueToDisplay($field, $values, array $options = [])
+    {
+        $color_aliases = [
+            'grey'      => 'gray',
+            'darkgrey'  => 'darkgray'
+        ];
+        $printer = new Printer();
+        if (strpos($field, '_virtual_toner') === 0) {
```

Comment:
You can use the PHP 8 function `str_starts_with`. Polyfill lib is loaded for PHP < 8.0.
```suggestion
        if (str_starts_with($field, '_virtual_toner')) {
```